doc/nfadmin.1
.TH NFADMIN 1 "August 2004" "Newts" "Newts Reference Manual"

.SH NAME
nfadmin - change notesfiles' director options

.SH SYNOPSIS
.B nfadmin
[\fIoptions\fR] \fINOTESFILE\fR...

.SH DESCRIPTION
.B nfadmin
displays and alters director settings for notesfiles in which you have director
permissions.

The command-line options for `nfadmin' are non-standard; they do not strictly
follow normal GNU usage.

.SH FLAG OPTIONS

Flag options set or clear boolean flags, such as "moderated", for notesfiles.
The short option version of flag options uses either a \fB+\fR or a \fB-\fR
appended to the end of the option (for example, \fB-a+\fR or \fB-M-\fR) to
activate or deactive the flag.  Alternately, flag options can be set using
standard long options; in this case, supply an argument of "yes", "no", "on",
or "off".

.TP
\fB\-a\fR, \fB\-\^\-anonymous\fR=\fIFLAG\fR
Turn on or off anonymous posting in the specified notesfiles.

.TP
\fB\-A\fR, \fB\-\^\-archive\fR=\fIFLAG\fR
Turn on or off archive status for the specified notesfiles.

.TP
\fB\-l\fR, \fB\-\^\-locked\fR=\fIFLAG\fR
Lock or unlock the specified notesfiles; only directors are permitted to enter a
locked notesfile.

.TP
\fB\-M\fR, \fB\-\^\-moderated\fR=\fIFLAG\fR
Turn on or off moderation for the specifed notesfiles.

.TP
\fB\-o\fR, \fB\-\^\-open\fR=\fIFLAG\fR
"Open" or "close" the specified notesfiles; this is the same as unlocking and
locking, respectively.

.SH VALUE OPTIONS

Value options allow you to specify particular values for various
notesfile settings.

.TP
\fB\-d\fR, \fB\-\^\-expire-dirmsg\fR=\fIVALUE\fR
Change the "expire by director message" value for the specified notesfiles.
This options controls how notes in the notesfile should be affected by their
director messages during archiving.  Valid values are "on", "off", "any",
"ignore", and "default".

.TP
\fB\-e\fR, \fB\-\^\-expire-time\fR=\fIVALUE\fR
Set the number of days before notes are archived in the specified notesfiles.
Valid values are "never", "default", and positive integers; 0 is considered the
same as "default".

.TP
\fB\-E\fR, \fB\-\^\-expire-action\fR=\fIVALUE\fR
Change the action to be performed on expired notes in the specified notesfiles.
Valid values are "archive", "delete", and "default".

.TP
\fB\-L\fR, \fB\-\^\-max-note-size\fR=\fIVALUE\fR
Change the maximum size for individual notes in the specified notesfiles.  A
value can be entered as a simple number, which will be interpreted as a number
of bytes, or in the form "23k", which will be interpreted as a number of
kilobytes.  The absolute maximum is 2929k.

.TP
\fB\-W\fR, \fB\-\^\-minimum-notes\fR=\fIVALUE\fR
Change the number of notes which will be left in the notesfile at minimum after
archiving.  Valid values are "default" and positive integers.

.SH OTHER OPTIONS

.TP
\fB\-h\fR, \fB\-\^\-help\fR
Print a summary of usage and command-line options for
.B nfadmin
and exit.

.TP
\fB\-\^\-debug\fR
Print debugging messages to standard error.

.TP
\fB\-\^\-version\fR
Print version information for
.B nfadmin
and exit.

.SH AUTHOR
.B Newts
was written and is maintained by Tyler Berry <tyler+newts@thoughtlocker.net>.

.SH SEE ALSO
\fBautoseq\fR(1), \fBchecknotes\fR(1), \fBgetnote\fR(1), \fBmknf\fR(1),
\fBnfdump\fR(1), \fBnfload\fR(1), \fBnfmail\fR(1), \fBnfpipe\fR(1),
\fBnfprint\fR(1), \fBnfstats\fR(1), \fBnftimestamp\fR(1), \fBnotes\fR(1),
\fBrmnf\fR(1)

The full documentation for
.B Newts
is maintained as a Texinfo manual.  If the
.B info
and
.B Newts
programs are properly installed at your site, the command
.IP
.B info newts
.PP
should give you access to the complete manual.


NEWS
Newts NEWS                                                         -*- Text -*-
-------------------------------------------------------------------------------

* ______ - Newts 0.15

- More unit tests.
- More changes and additions to the client API.
- More documentation of the client API.

* Wednesday, December 14, 2005 - Newts 0.14.8

- Include the SVN revision number in the output of --version.
- Added unit tests for the client API.
- Began adding documentation for the client API in a Doxygen-compatible way.
- Substantial changes and additions to the client API.
- Substantial refactoring in the main client, both to comply with the revised
  client API and to clean the code up in general.
- Modified the utility frontend programs to use the revised client API.
- Refactored the pager to be modular, for purposes of future rewrites to add
  things like Unicode support.
- Use gnulib-tool to generate Gnulib support.
- Automatically use the most recent version of Libtool.

* Tuesday, July 26, 2005 - Newts 0.14.7

- Fixed signal handling so that suspending a spawned process works correctly.
- Added David Mellis's python bindings to the repository and distribution.

* Friday, June 16, 2005 - Newts 0.14.6

- Fixed a bug in nfdump where the last note in a dumped notesfile was omitted.
- Added 'g' to perform one-touch seqing.

* Monday, March 21, 2005 - Newts 0.14.5

- Updated to Libtool 1.5.14.
- Fixed a segfault when trying to use the alternate sequencer.
- Made the alternate sequencer activate the sequencer automatically.
- Made the alternate sequencer work correctly with the '--user' option.
- Fixed the broken --time option, which was not actually setting the time at
  all. Oops.
- Fixed a few odd display bugs.
- Display messages are now 'Unicode-friendly'.
  See <http://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html> for details.
- Resynced with Gnulib.

* Sunday, January 23, 2005 - Newts 0.14.4

- Updated to Libtool 1.5.10.
- Corrected some mistakes in the Texinfo manual.
- Fixed a bug which would cause a segfault when replying to a note from a local
  user who had been deleted from the system, or forwarding a mail from a note
  written by such a user.
- Fixed a bug in nfload where the dumpfile would fail to open if nfload was run
  as root. (It was dropping root permissions and not picking them back up.)
- Resynced with Gnulib, which involved some fairly major updates to xmalloc and
  friends, as the interface Gnulib uses changed.

* Tuesday, September 14, 2004 - Newts 0.14.3

- Corrected a typo in nfadmin.1.
- Synchronized with new files in Gnulib.

* Tuesday, August 10, 2004 - Newts 0.14.2

- Corrected a Whole Bunch of inconsistencies between the Texinfo manual, the
  newly-developed manpages, and the actual operation of the programs.
- Added manpages for all currently-built programs.
- Renamed 'printnote' to 'getnote'.
- Removed the '-d' short option from all programs, which used to correspond to
  '--debug'.  It was blocking the use of '-d' for more important options which
  are far more likely to see use.

* Tuesday, June 29, 2004 - Newts 0.14.1

- Completely redid all compatibility code using a recent CVS version of
  Gnulib.

* Thursday, June 10, 2004 - Newts 0.14

- Adding the missing code to the nfload program which will load access entries
  from the dumpfile into the notesfile.
- Added debugging support to all of the various binaries built by the package,
  via the --debug command-line option.
- Added an option for debugging that allows the notes client to setuid to an
  arbitrary user when being run by root. Useful for running notes inside gdb.
- Changed the license for the Newts Reference Manual - it's no longer GFDL.
- Fixed a bug which would cause a segfault when reading a note from a local
  user who had been deleted from the system.
- Reorganized the entire directory structure of the source tree.
- Reorganized header files to install the relevant interfaces.
- Began systematic documentation of the libnewtsclient interface.

* Tuesday, March 9, 2004 - Newts 0.13.2

- Fixed a display bug which caused informational messages and prompts to fail
  to display when a person tried to enter a notesfile without rights.
- Syncronized with new files in Gnulib.

* Friday, January 16, 2003 - Newts 0.13.1

- Synchronized with new releases of GNU Autotools and Gnulib.

* Tuesday, December 9, 2003 - Newts 0.13

- Added nfdump, completing the set of basic utilities.

clients/frontends/nfadmin.c
/*
 * nfadmin.c - command-line interface to director options
 *
 * This file is part of the Newts notesfile system.
 * Copyright (C) 2003, 2004, 2005 Tyler Berry.
 *
 * Based in part on nfadmin.c from the UIUC notes distribution by Ray Essick
 * and Rob Kolstad.  Any work derived from this source code is required to
 * retain this notice.
 *
 * Newts is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * Newts is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Newts; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include "frontend.h"

#include "dirname.h"
#include "error.h"
#include "getopt.h"
#include "newts/uiuc.h"
#include "newts/uiuc-compatibility.h"

#if STDC_HEADERS
# include <limits.h>
#endif

static void summary (struct notesfile *nf);

/* Whether to display debugging messages. */
int debug = FALSE;

int
main (int argc, char **argv)
{
  List nflist;
  struct notesfile nf;
  int error_occurred = FALSE;

  int anonymous = 0;
  int archive = 0;
  int locked = 0;
  int moderated = 0;
  int expire_dirmsg = -2;
  int expire_time = -2;
  int expire_action = -2;
  int max_note_size = -2;
  int minimum_notes = -2;

  int i;

  int opt;
  int option_index = 0;

  struct option long_options[] =
    {
      {N_("anonymous"),2,0,'{'},
      {N_("archive"),2,0,'}'},
      {N_("debug"),0,0,'D'},
      {N_("locked"),2,0,'['},
      {N_("moderated"),2,0,']'},
      {N_("open"),2,0,'('},
      {N_("expire-dirmsg"),1,0,'d'},
      {N_("expire-time"),1,0,'e'},
      {N_("expire-action"),1,0,'E'},
      {N_("help"),0,0,'h'},
      {N_("max-note-size"),1,0,'L'},
      {N_("minimum-notes"),1,0,'W'},
      {N_("version"),0,0,0},
      {0,0,0,0}
    };

  memset (&nf, 0, sizeof (struct notesfile));

#ifdef __GLIBC__
  program_name = program_invocation_short_name;
#else
  program_name = base_name (argv[0]);
#endif

  /* Initialize i18n. */

#ifdef HAVE_SETLOCALE
  setlocale (LC_ALL, "");
#endif

#if ENABLE_NLS
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif

  setup ();

  /* Rewrite the command line options so that getopt can handle them. */

  for (i=1; i<argc; i++)
    {
      /* This won't catch compound options, but short of calling getopt twice
       * that's not going to happen anyway.
       */

      if (argv[i][0] == '-' &&    /* Only 'tweak' options. */
          (strcmp (argv[i], N_("-N")) == 0 || strcmp (argv[i], N_("-n")) == 0 ||
           strcmp (argv[i], N_("-Q")) == 0 || strcmp (argv[i], N_("-q")) == 0 ||
           strcmp (argv[i], N_("-P")) == 0 || strcmp (argv[i], N_("-p")) == 0 ||
           strcmp (argv[i], N_("-B")) == 0 || strcmp (argv[i], N_("-b")) == 0))
        {
          fprintf (stderr, _("%s: invalid option -- %c\n"), argv[0], argv[i][1]);
          fprintf (stderr, _("Try '%s --help' for more information.\n"),
                   program_name);
          exit (EXIT_FAILURE);
        }
      else if (strcmp (argv[i], N_("-a+")) == 0 || strcmp (argv[i], N_("-a")) == 0)
        {
          argv[i] = N_("-N");
        }
      else if (strcmp (argv[i], N_("-a-")) == 0)
        {
          argv[i] = N_("-n");
        }
      else if (strcmp (argv[i], N_("-M+")) == 0 || strcmp (argv[i], N_("-M")) == 0)
        {
          argv[i] = N_("-Q");
        }
      else if (strcmp (argv[i], N_("-M-")) == 0)
        {
          argv[i] = N_("-q");
        }
      else if (strcmp (argv[i], N_("-o-")) == 0 || strcmp (argv[i], N_("-l")) == 0 ||
               strcmp (argv[i], N_("-l+")) == 0)
        {
          argv[i] = N_("-P");
        }
      else if (strcmp (argv[i], N_("-o+")) == 0 || strcmp (argv[i], N_("-o")) == 0 ||
               strcmp (argv[i], N_("-l-")) == 0)
        {
          argv[i] = N_("-p");
        }
      else if (strcmp (argv[i], N_("-A+")) == 0 || strcmp (argv[i], N_("-A")) == 0)
        {
          argv[i] = N_("-B");
        }
      else if (strcmp (argv[i], N_("-A-")) == 0)
        {
          argv[i] = N_("-b");
        }
    }

  while ((opt = getopt_long (argc, argv, N_("bBd:e:E:hl:nNpPqQW:"),
                             long_options, &option_index)) != -1)
    {
      switch (opt)
        {
        case 0:
          {
            printf_version_string (N_("nfadmin"));

            teardown ();

            if (fclose (stdout) == EOF)
              error (EXIT_FAILURE, errno, _("error writing output"));
            exit (EXIT_SUCCESS);
          }

        case '{':  /* anonymous */
          if (optarg == NULL || strcmp (optarg, N_("yes")) == 0 ||
              strcmp (optarg, N_("on")) == 0 || strcmp (optarg, N_("y")) == 0 ||
              strcmp (optarg, N_("+")) == 0)
            anonymous = 1;
          else if (strcmp (optarg, N_("no")) == 0 || strcmp (optarg, N_("off")) == 0 ||
                   strcmp (optarg, N_("n")) == 0 || strcmp (optarg, N_("-")) == 0)
            anonymous = -1;
          else
            {
              fprintf (stderr, _("%s: invalid argument -- %s\n"),
                       argv[0], optarg);
              fprintf (stderr, _("Try '%s --help' for more information.\n"),
                       program_name);

              teardown ();

              exit (EXIT_FAILURE);
            }
          break;

        case '}':  /* archive */
          if (optarg == NULL || strcmp (optarg, N_("yes")) == 0 ||
              strcmp (optarg, N_("on")) == 0 || strcmp (optarg, N_("y")) == 0 ||
              strcmp (optarg, N_("+")) == 0)
            archive = 1;
          else if (strcmp (optarg, N_("no")) == 0 || strcmp (optarg, N_("off")) == 0 ||
                   strcmp (optarg, N_("n")) == 0 || strcmp (optarg, N_("-")) == 0)
            archive = -1;
          else
            {
              fprintf (stderr, _("%s: invalid argument -- %s\n"),
                       argv[0], optarg);
              fprintf (stderr, _("Try '%s --help' for more information.\n"),
                       program_name);

              teardown ();

              exit (EXIT_FAILURE);
            }
          break;

        case '[':  /* locked */
          if (optarg == NULL || strcmp (optarg, N_("yes")) == 0 ||
              strcmp (optarg, N_("on")) == 0 || strcmp (optarg, N_("y")) == 0 ||
              strcmp (optarg, N_("+")) == 0)
            locked = 1;
          else if (strcmp (optarg, N_("no")) == 0 || strcmp (optarg, N_("off")) == 0 ||
                   strcmp (optarg, N_("n")) == 0 || strcmp (optarg, N_("-")) == 0)
            locked = -1;
          else
            {
              fprintf (stderr, _("%s: invalid argument -- %s\n"),
                       argv[0], optarg);
              fprintf (stderr, _("Try '%s --help' for more information.\n"),
                       program_name);

              teardown ();

              exit (EXIT_FAILURE);
            }
          break;

        case '(':  /* open -- locked in reverse */
          if (optarg == NULL || strcmp (optarg, N_("yes")) == 0 ||
              strcmp (optarg, N_("on")) == 0 || strcmp (optarg, N_("y")) == 0 ||
              strcmp (optarg, N_("+")) == 0)
            locked = -1;
          else if (strcmp (optarg, N_("no")) == 0 || strcmp (optarg, N_("off")) == 0 ||
                   strcmp (optarg, N_("n")) == 0 || strcmp (optarg, N_("-")) == 0)
            locked = 1;
          else
            {
              fprintf (stderr, _("%s: invalid argument -- %s\n"),
                       argv[0], optarg);
              fprintf (stderr, _("Try '%s --help' for more information.\n"),
                       program_name);

              teardown ();

              exit (EXIT_FAILURE);
            }
          break;

        case ']':  /* moderated */
          if (optarg == NULL || strcmp (optarg, N_("yes")) == 0 ||
              strcmp (optarg, N_("on")) == 0 || strcmp (optarg, N_("y")) == 0 ||
              strcmp (optarg, N_("+")) == 0)
            moderated = 1;
          else if (strcmp (optarg, N_("no")) == 0 || strcmp (optarg, N_("off")) == 0 ||
                   strcmp (optarg, N_("n")) == 0 || strcmp (optarg, N_("-")) == 0)
            moderated = -1;
          else
            {
              fprintf (stderr, _("%s: invalid argument -- %s\n"),
                       argv[0], optarg);
              fprintf (stderr, _("Try '%s --help' for more information.\n"),
                       program_name);

              teardown ();

              exit (EXIT_FAILURE);
            }
          break;

        case 'b':
          archive = -1;
          break;

        case 'B':
          archive = 1;
          break;

        case 'D':
          debug = TRUE;
          break;

        case 'd':
          if (strcmp (optarg, N_("on")) == 0 || strcmp (optarg, N_("yes")) == 0)
            expire_dirmsg = DIRON;
          else if (strcmp (optarg, N_("off")) == 0 || strcmp (optarg, N_("no")) == 0)
            expire_dirmsg = DIROFF;
          else if (strncmp (optarg, N_("ign"), 3) == 0)
            expire_dirmsg = DIRNOCARE;
          else if (strncmp (optarg, N_("any"), 3) == 0)
            expire_dirmsg = DIRANYON;
          else if (strncmp (optarg, N_("def"), 3) == 0 ||
                   strncmp (optarg, N_("dfl"), 3) == 0)
            expire_dirmsg = DIRDFLT;
          else
            {
              fprintf (stderr, _("%s: invalid argument -- %s\n"),
                       argv[0], optarg);
              fprintf (stderr, _("Try '%s --help' for more information.\n"),
                       program_name);

              teardown ();

              exit (EXIT_FAILURE);
            }
          break;

        case 'e':
          {
            long temp;
        char *tail;

        if (strncmp (optarg, N_("nev"), 3) == 0 || strcmp (optarg, N_("n")) == 0)
              expire_time = -1;
        else if (strncmp (optarg, N_("def"), 3) == 0 ||
                 strcmp (optarg, N_("d")) == 0 ||
                 strncmp (optarg, N_("dfl"), 3) == 0)
              expire_time = 0;
            else
              {
                temp = strtol (optarg, &tail, 10);
                if (temp == 0 && optarg == tail)
                  {
                    fprintf (stderr, _("%s: invalid argument -- %s\n"),
                             argv[0], optarg);
                    fprintf (stderr, _("Try '%s --help' for more information.\n"),
                             program_name);

                    teardown ();

                    exit (EXIT_FAILURE);
                  }
                else if ((temp == LONG_MAX && errno == ERANGE) || temp > INT_MAX)
                  {
                    fprintf (stderr, _("%s: invalid argument -- %s (exceeded maximum value)\n"),
                             argv[0], optarg);
                    fprintf (stderr, _("Try '%s --help' for more information.\n"),
                             program_name);

                    teardown ();

                    exit (EXIT_FAILURE);
                  }
                else if (temp < 0)
                  {
                    fprintf (stderr, _("%s: invalid argument -- %s (no negative values)\n"),
                             argv[0], optarg);
                    fprintf (stderr, _("Try '%s --help' for more information.\n"),
                             program_name);

                    teardown ();

                    exit (EXIT_FAILURE);
                  }
                else
                  {
                    expire_time = temp;
                  }
              }

            break;
          }

        case 'E':
          if (strncmp (optarg, N_("arc"), 3) == 0)
            expire_action = KEEPYES;
          else if (strncmp (optarg, N_("del"), 3) == 0)
            expire_action = KEEPNO;
          else if (strncmp (optarg, N_("def"), 3) == 0 ||
                   strncmp (optarg, N_("dfl"), 3) == 0)
            expire_action = KEEPDFLT;
          else
            {
              fprintf (stderr, _("%s: invalid argument -- %s\n"),
                       argv[0], optarg);
              fprintf (stderr, _("Try '%s --help' for more information.\n"),
                       program_name);

              teardown ();

              exit (EXIT_FAILURE);
            }
          break;

        case 'f':
          break;

        case 'L':
          {
            long temp;
        char *tail;

            temp = strtol (optarg, &tail, 10);
            if (strcmp (tail, N_("k")) == 0 || strcmp (tail, N_("K")) == 0)
              temp *= 1024;
            if (temp == 0 && optarg == tail)
              {
                fprintf (stderr, _("%s: invalid argument -- %s\n"),
                         argv[0], optarg);
                fprintf (stderr, _("Try '%s --help' for more information.\n"),
                         program_name);

                teardown ();

                exit (EXIT_FAILURE);
              }
            else if ((temp == LONG_MAX && errno == ERANGE) || temp > HARDMAX)
              {
                fprintf (stderr, _("%s: invalid argument -- %s (exceeded maximum value)\n"),
                         argv[0], optarg);
                fprintf (stderr, _("Try '%s --help' for more information.\n"),
                         program_name);

                teardown ();

                exit (EXIT_FAILURE);
              }
            else if (temp < 0)
              {
                fprintf (stderr, _("%s: invalid argument -- %s (no negative values)\n"),
                         argv[0], optarg);
                fprintf (stderr, _("Try '%s --help' for more information.\n"),
                         program_name);

                teardown ();

                exit (EXIT_FAILURE);
              }
            else
              {
                max_note_size = temp;
              }

            break;
          }

        case 'n':
          anonymous = -1;
          break;

        case 'N':
          anonymous = 1;
          break;

        case 'p':
          locked = -1;
          break;

        case 'P':
          locked = 1;
          break;

        case 'q':
          moderated = -1;
          break;

        case 'Q':
          moderated = 1;
          break;

        case 'W':
          {
            long temp;
            char *tail;

            if (strncmp (optarg, N_("def"), 3) == 0 ||
                strcmp (optarg, N_("d")) == 0 ||
                strncmp (optarg, N_("dfl"), 3) == 0)
              expire_time = 0;
            else
              {
                temp = strtol (optarg, &tail, 10);
                if (temp == 0 && optarg == tail)
                  {
                    fprintf (stderr, _("%s: invalid argument -- %s\n"),
                             argv[0], optarg);
                    fprintf (stderr, _("Try '%s --help' for more information.\n"),
                             program_name);

                    teardown ();

                    exit (EXIT_FAILURE);
                  }
                else if ((temp == LONG_MAX && errno == ERANGE) || temp > INT_MAX)
                  {
                    fprintf (stderr, _("%s: invalid argument -- %s (exceeded maximum value)\n"),
                             argv[0], optarg);
                    fprintf (stderr, _("Try '%s --help' for more information.\n"),
                             program_name);

                    teardown ();

                    exit (EXIT_FAILURE);
                  }
                else if (temp < 0)
                  {
                    fprintf (stderr, _("%s: invalid argument -- %s (no negative values)\n"),
                             argv[0], optarg);
                    fprintf (stderr, _("Try '%s --help' for more information.\n"),
                             program_name);

                    teardown ();

                    exit (EXIT_FAILURE);
                  }
                else
                  {
                    minimum_notes = temp;
                  }
              }

            break;
          }

        case 'h':
          printf (_("Usage: %s [OPTION]... NOTESFILE...\n"
                    "Adjust director options for the specified NOTESFILE(s).\n\n"),
                  program_name);

          printf (_("General options:\n"
                    "  -h, --help                Display this help and exit\n"
                    "      --debug               Display debugging messages\n"
                    "      --version             Display version information and exit\n\n"));

          printf (_("Value options:\n"
                    "  -d, --expire-dirmsg=VAL   Set expiration by director message flag\n"
                    "  -e, --expire-time=VAL     Set number of days before note expiration\n"
                    "  -E, --expire-action=VAL   Set action to take upon expiration\n"
                    "  -L, --max-note-size=VAL   Set maximum note size in bytes or kilobytes (-k)\n"
                    "  -W, --minimum-notes=VAL   Set minimum notes to retain after archiving\n\n"));

          printf (_("Flag options:\n"
                    "  -a, --anonymous           Set value of 'anonymous' flag\n"
                    "  -A, --archive             Set value of 'archive' flag\n"
                    "  -l, --locked              Set value of 'locked' flag\n"
                    "  -M, --moderated           Set value of 'moderated' flag\n"
                    "  -o, --open                Set value of 'locked' flag backwards\n\n"));

          printf (_("Value options can be set to whatever the various options allow; for details,\n"
                    "see 'info nfadmin'. Flag options can be set with + or - following the short\n"
                    "options, or 'on', 'off', 'yes', and 'no' for the long options.\n\n"));

          printf (_("Report bugs to <%s>.\n"), PACKAGE_BUGREPORT);

          teardown ();

          if (fclose (stdout) == EOF)
            error (EXIT_FAILURE, errno, _("error writing output"));
          exit (EXIT_SUCCESS);

        case '?':
          fprintf (stderr, _("Try '%s --help' for more information.\n"),
                   program_name);

          teardown ();

          exit (EXIT_FAILURE);
        }
    }

  if (optind == argc)
    {
      fprintf (stderr, _("%s: too few arguments\n"), program_name);
      fprintf (stderr, _("Try '%s --help' for more information.\n"),
               program_name);

      teardown ();

      exit (EXIT_FAILURE);
    }

  list_init (&nflist,
             (void * (*) (void)) nfref_alloc,
             (void (*) (void *)) nfref_free,
             NULL);

  /* TRANSLATORS: Please keep this aligned as is; I know that's annoying but the
   * layout is fussy.
   */
  printf (_("        Notesfile         : Lock Anon Arch Modr Minm Expr Expr Expr   #    Maxm\n"
            "------------------------- : -------Flags------- Note Time Actn DirM Notes  NtSz\n"));

  while (optind < argc)
    parse_nf (argv[optind++], &nflist);

  {
    ListNode *node = list_head (&nflist);

    while (node != NULL)
      {
        int error;
        struct uiuc_opts *opts;

        error = open_nf ((newts_nfref *) list_data (node), &nf);

        node = list_next (node);

        if (error != NEWTS_NO_ERROR)
          {
            printf (_("%-25.25s : Error opening notesfile\n"),
                    nfref_pretty_name (nf.ref));
            error_occurred = TRUE;
            continue;
          }

        if (!allowed (&nf, DIRECTOR))
          {
            printf (_("%-25.25s : Not a director\n"),
                    nfref_pretty_name (nf.ref));
            continue;
          }

        opts = (struct uiuc_opts *) nf.opts;

        if (anonymous)
          {
            if (anonymous > 0)
              nf.options |= NF_ANONYMOUS;
            else
              nf.options &= ~NF_ANONYMOUS;
          }

        if (locked)
          {
            if (locked > 0)
              nf.options |= NF_LOCKED;
            else
              nf.options &= ~NF_LOCKED;
          }

        if (archive)
          {
            if (archive > 0)
              nf.options |= NF_ARCHIVE;
            else
              nf.options &= ~NF_ARCHIVE;
          }

        if (moderated)
          {
            if (moderated > 0)
              nf.options |= NF_MODERATED;
            else
              nf.options &= ~NF_MODERATED;
          }

        if (expire_action != -2)
          opts->expire_action = expire_action;

        if (expire_dirmsg != -2)
          opts->expire_by_dirmsg = expire_dirmsg;

        if (expire_time != -2)
          opts->expire_threshold = expire_time;

    if (minimum_notes != -2)
          opts->minimum_notes = minimum_notes;

        if (max_note_size != -2)
          opts->maximum_note_size = max_note_size;

        modify_nf (&nf);

        summary (&nf);
      }
  }

  teardown ();

  if (fclose (stdout) == EOF)
    error (EXIT_FAILURE, errno, _("error writing output"));

  exit (error_occurred ? EXIT_FAILURE : EXIT_SUCCESS);
}

static void
summary (struct notesfile *nf)
{
  struct uiuc_opts *opts = (struct uiuc_opts *) nf->opts;

  printf (N_("%-25.25s : "), nfref_pretty_name (nf->ref));
  printf (N_("%s %s %s %s "),
          nf->options & NF_LOCKED ? _("YES ") : _(" NO "),
          nf->options & NF_ANONYMOUS ? _("YES ") : _(" NO "),
          nf->options & NF_ARCHIVE ? _("YES ") : _(" NO "),
          nf->options & NF_MODERATED ? _("YES ") : _(" NO "));
  printf (N_("%4d "), opts->minimum_notes);
  if (opts->expire_threshold == -1)
    printf (_("NEVR "));
  else if (opts->expire_threshold == 0)
    printf (_("DFLT "));
  else
    printf (N_("%4d "), opts->expire_threshold);
  printf (N_("%s "), opts->expire_action == KEEPNO ? _("DEL ") :
          opts->expire_action == KEEPYES ? _("ARCH") : _("DFLT"));
  printf (N_("%s "), opts->expire_by_dirmsg == DIRON ? _(" ON ") :
          opts->expire_by_dirmsg == DIROFF ? _("OFF ") :
          opts->expire_by_dirmsg == DIRANYON ? _("ANY ") :
          opts->expire_by_dirmsg == DIRNOCARE ? _("IGNR") : _("DFLT"));
  printf (N_("%5d "), nf->total_notes);
  printf (N_("%4dk\n"), opts->maximum_note_size / 1024);

  return;
}
